Block ids pregenerate

wachterjohannes opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | #7672 
| Related issues/PRs | https://github.com/sulu/sulu/pull/8262
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR add the pregeneration of block-ids if the server respond blocks with missing id.

#### Why?

When blocks are created e.g. in fixtures the user could forget the id and so the UI adds them later.

wachterjohannes committed:
Pre-generate block IDs for blocks without IDs from server

When the server responds with block data that doesn't have IDs,
automatically generate IDs for those blocks using the generateBlockIds
function. This ensures all blocks have IDs even when loaded from legacy
data or external sources.

Implementation:
- Added ensureBlockIds method to detect and generate missing IDs
- Called in constructor for initial data
- Called in componentDidUpdate for prop changes
- Prevents concurrent ID generation with isGeneratingIds flag
- Validates generated IDs before applying them
- Only processes blocks that are actually missing IDs

Test Coverage:
- Added test for pre-generating IDs for all blocks without IDs
- Added test for not generating when all blocks have IDs
- Added test for mixed case (some with, some without IDs)
- All 60 tests passing

alexander-schranz commented:
@wachterjohannes rebase this one so we can review it.

wachterjohannes commented:
> @wachterjohannes rebase this one so we can review it.

done :)

alexander-schranz commented on src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js:
```diff
@@ -91,13 +92,72 @@ class BlockCollection<T: string, U: {_id?: string, type: T, ...}> extends React.
+        // 1. The value reference changed (not the one we just set)
+        // 2. There are actually blocks without IDs
+        // 3. We're not currently generating IDs
+        if (prevProps.value !== value && !this.isGeneratingIds) {
```
if the param is not defined we also don't need to call `this.ensureBlockIds();`

alexander-schranz commented on src/Sulu/Bundle/AdminBundle/Resources/js/components/BlockCollection/BlockCollection.js:
```diff
@@ -50,6 +50,7 @@ class BlockCollection<T: string, U: {_id?: string, type: T, ...}> extends React.
     @observable expandedBlocks: Array<boolean> = [];
     @observable selectedBlocks: Array<boolean> = [];
     @observable mode: BlockMode = 'sortable';
+    @observable isGeneratingIds: boolean = false;
```
This feels strange. Is that really required? If we add, paste ablock it should already have a block id and so hasBlocksWithoutIds should always be false.

And if the store is updated by the backend which can happen and not provide ids it should generate ids. SO I would remove this state as I think it would make issues if store is edited from outside.

alexander-schranz replied:
@wachterjohannes did you miss this comment?

wachterjohannes replied:
@alexander-schranz no i wanted to avoid that the generateIds run multiple times at the same time see: https://github.com/sulu/sulu/pull/8288/files#diff-67f069e752b8a5f4447294d36596f4596b5237278dd9c3de93d5ab79e297538dR124

wachterjohannes committed:
Address PR review feedback

- Check if generateBlockIds is defined before calling ensureBlockIds in componentDidUpdate
- Remove isGeneratingIds check from componentDidUpdate to allow external store updates
- The isGeneratingIds flag in ensureBlockIds still prevents concurrent async operations
- Add test case for blocks without IDs when generateBlockIds is not provided
- All 61 tests passing
